services/QuillLMS/engines/evidence/spec/controllers/evidence/opinion_controller_spec.rb

```
@@ -0,0 +1,83 @@
+require 'rails_helper'
+
+module Evidence
+  RSpec.describe(OpinionController, :type => :controller) do
+    before { @routes = Engine.routes }
+
+    describe '#fetch' do 
+      let(:example_error) { 'example_error' }
+      let(:example_rule_uid) { 123 }
+      let(:incoming_payload) do 
+        { 
+          'entry' => 'eat junk food at home.', 
+          'prompt_text' => 'Schools should not allow junk food to be sold on campus, but'
+        }
+      end
+
+      let(:oapi_response) do 
+        {
+          'oapi_error' => example_error,
+          'highlight' => [{
+            'type': 'response',
+            'text': 'someText',
+            'character': 0  
+          }]
+        }
+      end
+
+      let!(:rule) do 
+        create(:evidence_rule, uid: example_rule_uid)
+      end
+
+      before do 
+        # If you are unsure of the motivation behind the following line,
+        # the answers are here: https://github.com/rspec/rspec-mocks/issues/1079
+        OpinionFeedbackAssembler && stub_const(
+          'Evidence::OpinionFeedbackAssembler::OAPI_ERROR_TO_RULE_UID', 
+          { example_error => example_rule_uid }
+        )
+        allow(::HTTParty).to receive(:post).and_return(oapi_response)
+      end
+
+      it 'should return a valid json response' do 
+        post :fetch, params: incoming_payload, as: :json
+
+        expect(JSON.parse(response.body)).to eq({
+          'concept_uid' => 'ConceptUID',
+          'feedback' => nil,
+          'feedback_type' => 'opinion',
+          'optimal' => false,
+          'response_id' => '0',
+          'highlight' => [
+            { 'type' => 'response', 'text' => 'someText', 'character' => 0 }
+          ],
+          'labels' => '',
+          'rule_uid' => example_rule_uid.to_s
+        })
+      end
+
+      context 'Rule.feedback exists' do 
```

Let's add some tests for:
1. Opinion API raises a 500.
2. A key that isn't in the `OAPI_ERROR_TO_RULE_UID`.

I believe the payloads using the current code aren't what we'd want them to be.

---

While I initially planned to use NewRelic.notice_error for these cases, I ended up letting them fail 'hard'. What's the desired behavior, if not failing hard (5xx)? I don't see a lot of value in returning HTTP 2xx and using NewRelic.notice_error, since the go server could interpret that as a green light to use the Oapi response.

---

I agree that's a possible path, but I don't think that's what the code will do. That is why a test would be helpful for these two scenarios. 

1. For the 500 error, I believe that `HTTParty` swallows 500 errors (e.g. you need to check if response was a `.success?` or use its `.code` method).
2. For the missing uid, you are essentially swallowing this error with `nil` checks. That works but I don't think the response is what we want, i.e. `optimal: nil`.